### ISSUE TITLE
Is true attention (full-context) calculation needed during training?

### ISSUE THREAD

**Issue by Andrey36652** [epfml/landmark-attention#2]
Thanks for your work!
While I understand how you reduce compute requirements during inference (by picking only k blocks with top softmax scores for their landmark tokens) it is not clear for me is the same true for training? Do we need to backpropagate through all blocks?

**Comment by mkrima** [comment]
Yes, the training does not use a cache and the top-k retrieval mechanism is not applied there. Instead all the blocks are in the context and full-context attention is computed. However, note that **the training can be done with a small context length** regardless of the inference context length. In the paper we use 512 training context length for our experiments (including when fine-tuning LLaMA).

**Comment by Andrey36652** [comment]
@mkrima Does it work because you change positional information of fetched blocks? For fetched blocks you preserve their relative ordering, but not absolute positions (in "true" context, for example, tokens' index range 0-32768), right? And fetched blocks end up being placed "inside" 512 training context and having according (new, in range 0-512) positional info?
I mean RoPE don't allow context extrapolation (as ALiBi for example).

**Comment by mkrima** [comment]
Yes.
